### PR TITLE
samples: serial_lte_modem add MQTT service

### PR DESCRIPTION
**Pull request by lats1980**
Add MQTT service to LTE serial modem.
Below proprietary MQTT AT commands are used:

- AT#XMQTTCON
- AT#XMQTTPUB
- AT#XMQTTSUB

Signed-off-by: Larry Tsai <larry.tsai@nordicsemi.no>

### CODE REVIEW

**Comment by lats1980** [comment]
@junqingzou , please review the MQTT service in SLM.

File: samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c [hunk 1]
```diff
@@ -0,0 +1,835 @@
...
+	}
+
+	ret = slm_util_hex_check(payload_buf,
+				evt->param.publish.message.payload.len);
+	if (ret) {
+		char *data_hex;
+		int size;
+
+		size = evt->param.publish.message.payload.len * 2;
+		data_hex = k_malloc(size);
```
Comment (lemrey) [inline samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c:150]: Should add proper error checking here.

File: samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c [hunk 2]
```diff
@@ -0,0 +1,835 @@
...
+			return err;
+		}
+		err = at_params_short_get(&at_param_list, 5, &retain);
+		if (err < 0) {
+			return err;
+		}
+		if (datatype == DATATYPE_HEXADECIMAL) {
+			u8_t *data_hex;
+
+			data_hex = k_malloc(msg_sz / 2);
```
Comment (lemrey) [inline samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c:707]: And here too

File: samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c [hunk 3]
```diff
@@ -0,0 +1,835 @@
...
+		break;
+	}
+}
+
+static void mqtt_thread_fn(void *arg1, void *arg2, void *arg3)
+{
+	int err;
+
+	while (1) {
+		if (client.internal.state < MQTT_STATE_TCP_CONNECTED) {
```
Comment (rlubos) [inline samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c:303]: The application should not use internal fields of the client structure (especiialy that it requires hacky solutions like copying the content of internal header). There's really no need for this.

The thread could be simply spawned during the initialization phase, and then we could have something like:
```
static K_SEM_DEFINE(socket_ready, 0, 1);
static bool mqtt_running;

...
static void mqtt_thread_fn(void *arg1, void *arg2, void *arg3) {
	while (true) {
		k_sem_take(&socket_ready, K_FOREVER);

		while (mqtt_running) {
				/* Do all the processing. Any error reported by mqtt_ functions or poll events should break the loop. */
		}
	}
}
```

`do_mqtt_connect` shall call `k_sem_give(&socket_ready);` as it does so far plus set the `mqtt_running = true;`.
`do_mqtt_disconnect` shall set `mqtt_running = false;` to break the loop (verified also after poll() call).

File: samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c (hunk 1 above)
Comment (lats1980) [reply samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c:132]: Error checking is added.

File: samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c (hunk 2 above)
Comment (lats1980) [reply samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c:693]: Error checking is added.

File: samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c [hunk 4]
```diff
@@ -0,0 +1,835 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
```
[the remark this answers is not in the document]
Comment (lats1980) [reply samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c]: Fixed.

File: samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.h [hunk 5]
```diff
@@ -0,0 +1,49 @@
+/*
+ * Copyright (c) 2019 Nordic Semiconductor ASA
```
[the remark this answers is not in the document]
Comment (lats1980) [reply samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.h]: Fixed.

File: samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c [hunk 6]
```diff
@@ -0,0 +1,835 @@
...
+
+	if (client.internal.state != MQTT_STATE_IDLE) {
+		return -EINPROGRESS;
+	}
+
+	client_init();
+
+	/* Start MQTT listening thread */
+	if (mqtt_thread_id != NULL) {
+		k_thread_resume(mqtt_thread_id);
```
[the remark this answers is not in the document]
Comment (lats1980) [reply samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c]: Removed.

File: samples/nrf9160/serial_lte_modem/README.rst [hunk 7]
```diff
@@ -141,6 +141,15 @@ The following proprietary GPS AT commands are used in this sample:
...
 * AT#XGPS=<op>[,<mask>]
 
+MQTT AT Commands
+****************
+
+The following proprietary MQTT AT commands are used in this sample:
+
+* AT#XMQTTCON=<op>[,<url>,<port>[,<sec_tag>]]
+* AT#XMQTTPUB=<topic>,<datatype>,<msg>,<qos>,<retain>
+* AT#XMQTTSUB=<topic>,<op>[,<qos>]
```
[the remark this answers is not in the document]
Comment (lats1980) [reply samples/nrf9160/serial_lte_modem/README.rst:151]: Descriptions on #XMQTTEVT and #XMQTTMSG are added to the document.

File: samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c (hunk 3 above)
Comment (lats1980) [reply samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c]: I've implement the above logic. ctx.connected is used instead of mqtt_running.

**Comment by lats1980** [comment]
> The language for the doc update is fine.
> I can't judge if the commands are listed in the correct place.

I've added some more descriptions on the document for the notification. Please review, thanks a lot!

File: samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c [hunk 8]
```diff
@@ -0,0 +1,824 @@
...
+	int err = -EINVAL;
+
+	if (ctx.connected) {
+		return -EINPROGRESS;
+	}
+
+	client_init();
+
+	/* Start MQTT listening thread */
+	mqtt_thread_id = k_thread_create(&mqtt_thread,
```
Comment (rlubos) [inline samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c:455]: The thread should be spawned once, not at each connect call. Should be moved to `slm_at_mqtt_init` (or created statically with K_THREAD_DEFINE macro, https://docs.zephyrproject.org/latest/reference/kernel/threads/index.html#spawning-a-thread)

File: samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c [hunk 9]
```diff
@@ -0,0 +1,824 @@
...
+{
+	int err;
+
+	while (1) {
+		/* Don't go any further until MQTT is connected */
+		k_sem_take(&mqtt_connected, K_FOREVER);
+		while (ctx.connected) {
+			if (fds.fd == INVALID_FDS) {
+				LOG_WRN("ERROR: no socket to poll");
+				continue;
```
Comment (rlubos) [inline samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c:297]: The internal loop should break in case of error, there's no point continuing - the client should reconnect.

File: samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c [hunk 10]
```diff
@@ -0,0 +1,824 @@
...
+			}
+			if ((fds.revents & POLLIN) == POLLIN) {
+				err = mqtt_input(&client);
+				if (err != 0) {
+					LOG_ERR("ERROR: mqtt_input %d", err);
+					continue;
+				}
+			}
+			if ((fds.revents & POLLERR) == POLLERR) {
+				LOG_ERR("POLLERR");
```
Comment (rlubos) [inline samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c:317]: Errors detected at poll level (POLLERR, POLNVAL) should shutdown the mqtt context manually (`mqtt_disconnect` or `mqtt_abort`, the latter is likely a better fit here) - in case of such errors, `mqtt_input` will not be called, so the mqtt engine will not be able to shut down on its own. Otherwise, we'll miss the `MQTT_EVT_DISCONNECT` event.

File: samples/nrf9160/serial_lte_modem/README.rst [hunk 11]
```diff
@@ -141,6 +141,22 @@ The following proprietary GPS AT commands are used in this sample:
...
+
+The following unsolicited notification reports the MQTT events.
```
Comment (ru-fu) [inline samples/nrf9160/serial_lte_modem/README.rst:153]: This is hard to understand. Maybe:
```suggestion
The following unsolicited notification indicates that an MQTT event occurred:
```

File: samples/nrf9160/serial_lte_modem/README.rst [hunk 12]
```diff
@@ -141,6 +141,22 @@ The following proprietary GPS AT commands are used in this sample:
...
+* #XMQTTEVT=<type>,<result>
+
+The following unsolicited notification reports the received publish message. The topic and message
+will then be reported.
```
Comment (ru-fu) [inline samples/nrf9160/serial_lte_modem/README.rst:157]:
```suggestion
The following unsolicited notification indicates that a publish message was received and reports the topic and the message:
```

And does it report the topic and the message, or the topic length and the message length?

File: samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c [hunk 13]
```diff
@@ -0,0 +1,824 @@
...
+			ctx.connected = false;
+			break;
+		}
+		LOG_DBG("MQTT client connected!");
+		break;
+
+	case MQTT_EVT_DISCONNECT:
+		if (evt->result != 0) {
+			LOG_ERR("MQTT disconnect failed %d", evt->result);
+			break;
```
Comment (junqingzou) [inline samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c:199]: could you rephrase the code to avoid two breaks for one case?

File: samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c (hunk 9 above)
Comment (junqingzou) [reply samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c:297]: ctx.connected means that mqtt_connect() is successful, which in turn means the fds.fd is not NULL. So I think this check is redundant

File: samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c [hunk 14]
```diff
@@ -0,0 +1,824 @@
...
+		LOG_ERR("ERROR: getaddrinfo failed %d", err);
+
+		return;
+	}
+
+	addr = result;
+	err = -ENOENT;
+
+	/* Look for address of the broker. */
+	while (addr != NULL) {
```
Comment (junqingzou) [inline samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c:352]: Possible to get multiple addrinfo in result? In other modules, always assume the first and only addrinfo in the result.

File: samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c [hunk 15]
```diff
@@ -0,0 +1,824 @@
...
+	}
+
+	err = fds_init(&client);
+	if (err != 0) {
+		LOG_ERR("ERROR: fds_init %d", err);
+		return err;
+	}
+
+	/** start polling now for CONNACK */
+	if (err == 0) {
```
Comment (junqingzou) [inline samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c:474]: Yet another redundant if

File: samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c [hunk 16]
```diff
@@ -0,0 +1,824 @@
...
+{
+	int err = -EINVAL;
+
+	u16_t op;
+	size_t url_sz = MQTT_MAX_URL_LEN;
+	size_t cid_sz = MQTT_MAX_CID_LEN;
+
+	switch (cmd_type) {
+	case AT_CMD_TYPE_SET_COMMAND:
+		if (at_params_valid_count_get(&at_param_list) == 0) {
```
Comment (junqingzou) [inline samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c:569]: No, the count starts from AT command itself, which is index 0, <op> is index 1. So the check should be count > 1

File: samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c [hunk 17]
```diff
@@ -0,0 +1,824 @@
...
+			return -EINVAL;
+		}
+		err = at_params_short_get(&at_param_list, 1, &op);
+		if (err < 0) {
+			return err;
+		}
+		if (op == 1) {
```
Comment (junqingzou) [inline samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c:576]: Please use AT_MQTTCON_CONNECT defined above

File: samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c [hunk 18]
```diff
@@ -0,0 +1,824 @@
...
+				err = at_params_int_get(&at_param_list, 5,
+							&ctx.sec_tag);
+				if (err < 0) {
+					return err;
+				}
+				ctx.sec_transport = true;
+			}
+			err = do_mqtt_connect();
+			break;
+		} else if (op == 0) {
```
Comment (junqingzou) [inline samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c:612]: Please use AT_MQTTCON_DISCONNECT defined above

File: samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c [hunk 19]
```diff
@@ -0,0 +1,824 @@
...
+							ctx.url, &url_sz);
+			if (err < 0) {
+				return err;
+			}
+			ctx.url[url_sz] = '\0';
+			err = at_params_int_get(&at_param_list, 4, &ctx.port);
+			if (err < 0) {
+				return err;
+			}
+			if (at_params_valid_count_get(&at_param_list) == 6) {
```
Comment (junqingzou) [inline samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c:602]: I recommend if count > 5, == is more strict

File: samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c [hunk 20]
```diff
@@ -0,0 +1,824 @@
...
+				if (err < 0) {
+					return err;
+				}
+				ctx.sec_transport = true;
+			}
+			err = do_mqtt_connect();
+			break;
+		} else if (op == 0) {
+			LOG_DBG("Disconnect from broker");
+			err = do_mqtt_disconnect();
```
Comment (junqingzou) [inline samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c:614]: There is a need to check ctx.connected in case of not connected yet

File: samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c [hunk 21]
```diff
@@ -0,0 +1,824 @@
...
+		err = mqtt_subscribe(&client, &subscription_list);
+	} else if (op == 0) {
+		err = mqtt_unsubscribe(&client, &subscription_list);
+	}
+
+	return err;
+}
+
+/**@brief handle AT#XMQTTCON commands
+ *  AT#XMQTTCON=<op>[,<url>,<port>[,<sec_tag>]]
```
Comment (junqingzou) [inline samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c:555]: Missing <client_id>? <op>[,<cid>,<url>,<port>[,<sec_tag>]]?

File: samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c [hunk 22]
```diff
@@ -0,0 +1,824 @@
...
+
+	u16_t qos, retain, datatype;
+	u8_t topic[MQTT_MAX_TOPIC_LEN];
+	size_t topic_sz = MQTT_MAX_TOPIC_LEN;
+	u8_t msg[MQTT_MESSAGE_BUFFER_LEN];
+	size_t msg_sz = MQTT_MESSAGE_BUFFER_LEN;
+
+	switch (cmd_type) {
+	case AT_CMD_TYPE_SET_COMMAND:
+		if (at_params_valid_count_get(&at_param_list) != 6) {
```
Comment (junqingzou) [inline samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c:665]: Again, I recommend if > 5

File: samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c [hunk 23]
```diff
@@ -0,0 +1,824 @@
...
+int slm_at_mqtt_init(void)
+{
+	return 0;
+}
+
+int slm_at_mqtt_uninit(void)
+{
+	client.broker = NULL;
+	fds.fd = INVALID_FDS;
+
```
Comment (junqingzou) [inline samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c:822]: The thread needs to be terminated here

File: samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c [hunk 24]
```diff
@@ -0,0 +1,824 @@
...
+{
+	int err;
+
+	err = mqtt_disconnect(&client);
+	if (err) {
+		LOG_ERR("ERROR: mqtt_disconnect %d", err);
+	}
+	slm_at_mqtt_uninit();
+
+	return err;
```
Comment (junqingzou) [inline samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c:492]: I guess that you need to suspend the thread here.
In connection, check the thread_id, if already created, do resume.
This is the key point o support re-connect to Broker

File: samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c (hunk 14 above)
Comment (rlubos) [reply samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c:352]: Theoretically `getaddrinfo` can return multiple results (mostly if both are allowed IPv4 and IPv6). If only IPv4 is requested, it's fine to avoid this check.

File: samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c (hunk 24 above)
Comment (rlubos) [reply samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c:492]: The thread will bo blocked on a semaphore waiting for another connection, thus making it inactive. No need to suspend explicitly.

File: samples/nrf9160/serial_lte_modem/README.rst (hunk 7 above)
[the remark this answers is not in the document]
Comment (lats1980) [reply samples/nrf9160/serial_lte_modem/README.rst]: > How about AT#XMQTTSUB=[,] and AT#XMQTTUNSUB=, as SUBSCRIBE and UNSUBSCRIBE are standard MQTT control packets and commonly used, so it would be good to have two explicit commands

AT#XMQTTSUB is implemented.
Syntax of subscribe/unsubscribe are:
* AT#XMQTTSUB=<topic>,<qos>
* AT#XMQTTUNSUB=<topic>

File: samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c (hunk 8 above)
Comment (lats1980) [reply samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c]: Now the thread is statically defined in K_THREAD_DEFINE.


K_THREAD_DEFINE(mqtt_thread, K_THREAD_STACK_SIZEOF(mqtt_thread_stack),
		mqtt_thread_fn, NULL, NULL, NULL,
		THREAD_PRIORITY, 0, K_NO_WAIT);

File: samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c (hunk 9 above)
Comment (lats1980) [reply samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c]: Yes, it is redundant. I've removed the check.

File: samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c (hunk 10 above)
Comment (lats1980) [reply samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c:276]: mqtt_abort is added to the poll error casees. MQTT_EVT_DISCONNECT is notified after calling mqtt_abort.

File: samples/nrf9160/serial_lte_modem/README.rst (hunk 11 above)
Comment (lats1980) [reply samples/nrf9160/serial_lte_modem/README.rst]: Thank you. I've modified the description as:

The following unsolicited notification indicates that an MQTT event occurred:

* #XMQTTEVT=<type>,<result>

File: samples/nrf9160/serial_lte_modem/README.rst (hunk 12 above)
Comment (lats1980) [reply samples/nrf9160/serial_lte_modem/README.rst]: The notification will first notify the data type, topic length and message length. Then the topic and message is notified. Carriage Return and Line Feed is inserted before topic and message. The description becomes:

The following unsolicited notification indicates that a publish message was
received and reports the topic and the message:

* #XMQTTMSG=<datatype>,<topic_length>,<message_length><CR><LF><topic><CR><LF><message>

File: samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c [hunk 25]
```diff
@@ -0,0 +1,824 @@
...
+	.fd = INVALID_FDS,
+};
+
+/**@brief Function to read the published payload.
+ */
+static int publish_get_payload(struct mqtt_client *c, size_t length)
+{
+	if (length > sizeof(payload_buf)) {
+		return -EMSGSIZE;
+	}
```
[the remark this answers is not in the document]
Comment (lats1980) [reply samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c:99]: As we discussed, I use predefined static array to avoid mem heap issue.

File: samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.h [hunk 26]
```diff
@@ -0,0 +1,49 @@
...
+ * @brief Uninitialize MQTT AT command parser.
+ *
+ * @retval 0 If the operation was successful.
+ *           Otherwise, a (negative) error code is returned.
+ */
+int slm_at_mqtt_uninit(void);
+
+/** @} */
+
+#endif /* SLM_AT_GPS_ */
```
[the remark this answers is not in the document]
Comment (lats1980) [reply samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.h:55]: slm_at_mqtt_clac added.

File: samples/nrf9160/serial_lte_modem/src/slm_at_host.c [hunk 27]
```diff
@@ -523,6 +533,11 @@ int slm_at_host_init(void)
 		LOG_ERR("GPS could not be initialized: %d", err);
 		return -EFAULT;
 	}
+	err = slm_at_mqtt_init();
+	if (err) {
+		LOG_ERR("MQTT could not be initialized: %d", err);
+		return -EFAULT;
+	}
```
[the remark this answers is not in the document]
Comment (lats1980) [reply samples/nrf9160/serial_lte_modem/src/slm_at_host.c:541]: slm_at_mqtt_clac added.

File: samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c [hunk 28]
```diff
@@ -0,0 +1,824 @@
...
+#define THREAD_PRIORITY		K_LOWEST_APPLICATION_THREAD_PRIO
+
+static struct k_thread mqtt_thread;
+static k_tid_t mqtt_thread_id;
+static K_THREAD_STACK_DEFINE(mqtt_thread_stack, THREAD_STACK_SIZE);
+
+/* Buffers for MQTT client. */
+static u8_t rx_buffer[MQTT_MESSAGE_BUFFER_LEN];
+static u8_t tx_buffer[MQTT_MESSAGE_BUFFER_LEN];
+static u8_t payload_buf[MQTT_MESSAGE_BUFFER_LEN];
```
[the remark this answers is not in the document]
Comment (lats1980) [reply samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c:77]: Still use static arrary to avoid mem heap issue.

File: samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c [hunk 29]
```diff
@@ -0,0 +1,824 @@
...
+static void handle_mqtt_publish_evt(struct mqtt_client *const c,
+					const struct mqtt_evt *evt)
+{
+	int ret;
+
+	ret = publish_get_payload(c,
+				evt->param.publish.message.payload.len);
+	if (ret < 0) {
+		LOG_ERR("mqtt_read_publish_payload: Failed! %d\n", ret);
+		sprintf(rsp_buf, "#XMQTTMSG: %d,0,0\r\n", ret);
```
[the remark this answers is not in the document]
Comment (lats1980) [reply samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c]: Yes, sending back error code by XMQTTMSG will mislead user. Now I use the #XMQTTEVT to send back error as other event.

#XMQTTEVT=<type>,<result>

File: samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c (hunk 13 above)
Comment (lats1980) [reply samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c]: Rephrased.

File: samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c (hunk 15 above)
Comment (lats1980) [reply samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c]: Removed.

File: samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c (hunk 16 above)
Comment (lats1980) [reply samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c]: As discussed, keep using strict check.

File: samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c [hunk 30]
```diff
@@ -0,0 +1,824 @@
...
+#define INVALID_FDS -1
+
+/**@brief MQTT connect operations. */
+enum slm_mqttcon_operation {
+	AT_MQTTCON_DISCONNECT,
+	AT_MQTTCON_CONNECT
+};
+
+/**@brief MQTT publish operations. */
+enum slm_mqttpub_operation {
```
[the remark this answers is not in the document]
Comment (lats1980) [reply samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c]: Removed.

File: samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c (hunk 18 above)
Comment (lats1980) [reply samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c]: Fixed.

File: samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c (hunk 19 above)
Comment (lats1980) [reply samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c:554]: Keep using strict check.

File: samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c (hunk 20 above)
Comment (lats1980) [reply samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c:569]: Added.

File: samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c (hunk 21 above)
Comment (lats1980) [reply samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c]: Added.

File: samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c (hunk 23 above)
Comment (lats1980) [reply samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c:810]: As Robert said, the thread is blocked so no need to be terminated.

File: samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c (hunk 17 above)
Comment (lats1980) [reply samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c]: Done.

File: samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c (hunk 14 above)
Comment (lats1980) [reply samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c:313]: I've modified the code here to check both IPv4 and IPv6 address.

File: samples/nrf9160/serial_lte_modem/README.rst [hunk 31]
```diff
@@ -146,6 +146,25 @@ Default SUPL server is as below:
 CONFIG_SLM_SUPL_SERVER="supl.google.com"
 CONFIG_SLM_SUPL_PORT=7276
 
+MQTT AT Commands
```
[the remark this answers is not in the document]
Comment (lats1980) [reply samples/nrf9160/serial_lte_modem/README.rst:150]: Done.

File: samples/nrf9160/serial_lte_modem/prj.conf [hunk 32]
```diff
@@ -55,6 +55,9 @@ CONFIG_DEVICE_POWER_MANAGEMENT=y
 
 # Enable SUPL client support
 #CONFIG_SUPL_CLIENT_LIB=y
+# MQTT
+CONFIG_MQTT_LIB=y
+CONFIG_MQTT_LIB_TLS=y
```
[the remark this answers is not in the document]
Comment (lats1980) [reply samples/nrf9160/serial_lte_modem/prj.conf]: Both CONFIG_MQTT_LIB and CONFIG_MQTT_LIB_TLS are added to Kconfig.

File: samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c [hunk 33]
```diff
@@ -0,0 +1,831 @@
...
+	param.message.topic.topic.size = topic_len;
+	param.message.payload.data = msg;
+	param.message.payload.len = msg_len;
+	param.message_id = sys_rand32_get();
+	param.dup_flag = 0;
+
+	return mqtt_publish(&client, &param);
+}
+
+static int do_mqtt_subscribe(u16_t op,
```
[the remark this answers is not in the document]
Comment (lats1980) [reply samples/nrf9160/serial_lte_modem/src/slm_at_mqtt.c:496]: Keep  internal function for both subscribe/unsubscribe to reduce code.
